### ISSUE TITLE
sql: can't use subquery results for span generation

### ISSUE THREAD

**Issue by turbo** [cockroachdb/cockroach#24282]
### Logs

```text
ssh root@db1.[REDACTED].net grep -F '[config]' cockroach-data/logs/cockroach.log
I180328 12:45:09.297723 1 util/log/clog.go:1085  [config] file created at: 2018/03/28 12:45:09
I180328 12:45:09.297723 1 util/log/clog.go:1085  [config] running on machine: db1
I180328 12:45:09.297723 1 util/log/clog.go:1085  [config] binary: CockroachDB CCL v2.0-beta.20180319 (linux amd64, built 2018/03/19 18:57:22, go1.10)
I180328 12:45:09.297723 1 util/log/clog.go:1085  [config] arguments: [cockroach start --certs-dir=certs --advertise-host=db1.[REDACTED].net --join=db1.[REDACTED].net:26257,db2.[REDACTED].net:26257,db3.[REDACTED].net:26257 --cache=25% --max-sql-memory=25%]
```

```text
ssh root@db2.[REDACTED].net grep -F '[config]' cockroach-data/logs/cockroach.log
I180328 12:46:51.699883 1 util/log/clog.go:1085  [config] file created at: 2018/03/28 12:46:51
I180328 12:46:51.699883 1 util/log/clog.go:1085  [config] running on machine: db2
I180328 12:46:51.699883 1 util/log/clog.go:1085  [config] binary: CockroachDB CCL v2.0-beta.20180319 (linux amd64, built 2018/03/19 18:57:22, go1.10)
I180328 12:46:51.699883 1 util/log/clog.go:1085  [config] arguments: [cockroach start --certs-dir=certs --advertise-host=db2.[REDACTED].net --join=db1.[REDACTED].net:26257,db2.[REDACTED].net:26257,db3.[REDACTED].net:26257 --cache=25% --max-sql-memory=25%]
```

```text
ssh root@db3.[REDACTED].net grep -F '[config]' cockroach-data/logs/cockroach.log
I180328 12:57:26.061779 1 util/log/clog.go:1085  [config] file created at: 2018/03/28 12:57:26
I180328 12:57:26.061779 1 util/log/clog.go:1085  [config] running on machine: db3
I180328 12:57:26.061779 1 util/log/clog.go:1085  [config] binary: CockroachDB CCL v2.0-beta.20180319 (linux amd64, built 2018/03/19 18:57:22, go1.10)
I180328 12:57:26.061779 1 util/log/clog.go:1085  [config] arguments: [cockroach start --certs-dir=certs --advertise-host=db3.[REDACTED].net --join=db1.[REDACTED].net:26257,db2.[REDACTED].net:26257,db3.[REDACTED].net:26257 --cache=25% --max-sql-memory=25%]
```

### Tables

```text
show create table fs_test3;
+----------+-----------------------------------------------------------------------------------------------------------------------------------+
|  Table   |                                                            CreateTable                                                            |
+----------+-----------------------------------------------------------------------------------------------------------------------------------+
| fs_test3 | CREATE TABLE fs_test3 (␤                                                                                                          |
|          |     id INTEGER NOT NULL,␤                                                                                                         |
|          |     "name" STRING NOT NULL,␤                                                                                                      |
|          |     metaphone_a STRING NOT NULL,␤                                                                                                 |
|          |     metaphone_b STRING NOT NULL,␤                                                                                                 |
|          |     last_change INTEGER NOT NULL,␤                                                                                                |
|          |     last_access INTEGER NOT NULL,␤                                                                                                |
|          |     size INTEGER NOT NULL,␤                                                                                                       |
|          |     misc STRING NOT NULL,␤                                                                                                        |
|          |     parent_id INTEGER NULL,␤                                                                                                      |
|          |     CONSTRAINT "primary" PRIMARY KEY (id ASC),␤                                                                                   |
|          |     INDEX fs_test3_name_idx ("name" ASC),␤                                                                                        |
|          |     INDEX fs_test3_metaphone_a_idx (metaphone_a ASC),␤                                                                            |
|          |     INDEX fs_test3_metaphone_b_idx (metaphone_b ASC),␤                                                                            |
|          |     INDEX fs_test3_parent_id_idx (parent_id ASC),␤                                                                                |
|          |     INDEX fs_test3_name_metaphone_a_metaphone_b_last_access_idx ("name" ASC, metaphone_a ASC, metaphone_b ASC, last_access ASC),␤ |
|          |     FAMILY "primary" (id, "name", metaphone_a, metaphone_b, last_change, last_access, size, misc, parent_id)␤                     |
|          | )                                                                                                                                 |
+----------+-----------------------------------------------------------------------------------------------------------------------------------+
```

```text
show create table fs_test3_cache_matpath;
+------------------------+-------------------------------------------------+
|         Table          |                   CreateTable                   |
+------------------------+-------------------------------------------------+
| fs_test3_cache_matpath | CREATE TABLE fs_test3_cache_matpath (␤          |
|                        |     id INT NOT NULL,␤                           |
|                        |     matpath INT[] NULL,␤                        |
|                        |     CONSTRAINT "primary" PRIMARY KEY (id ASC),␤ |
|                        |     FAMILY "primary" (id, matpath)␤             |
|                        | )                                               |
+------------------------+-------------------------------------------------+
```

### Problem

The actual problem I'm trying to solve is this: Given a list of IDs, resolve the materialized path of that entry. I would usually use a recursive CTE for this, but CDB doesn't support that.

So I'm caching the matpath in a separate table. This still leaves the names to resolve. I currently solve this client side:

1. Get the list of IDs
2. For each ID (in parallel), get the matpath
3. For each matpath, resolve the `name` column of each id (in parallel)

This is reasonably fast. Today @knz suggested to solve the last part in SQL using `unnest`. However, that is terribly slow. A comparison:

```sql
SELECT id, name 
FROM fs_test3 
WHERE id IN (
    SELECT * 
    FROM unnest((
        SELECT matpath 
        FROM fs_test3_cache_matpath 
        LIMIT 1
    ))
) ORDER BY id ASC;
```

takes 22 seconds (!). While

```sql
SELECT matpath 
        FROM fs_test3_cache_matpath 
        LIMIT 1;

-- ... then, 9,5027,5069 being the result

SELECT id, name 
FROM fs_test3 
WHERE id IN (9,5027,5069) ORDER BY id ASC;
```

takes about 90ms combined.

**Comment by jordanlewis** [comment]
Hi @turbo,

Unfortunately, we don't yet support efficient index selection based on the results from a subquery. What this means is that even though the subquery returns a concrete list of values, our optimizer doesn't know how to use those values to create efficient search spans in the database. As a result, the outer query uses a full table scan to search for the output.

Improving this is on our roadmap. The optimizer team might have more to say on this - cc @RaduBerinde.

As a workaround, I'm afraid the best advice we have at the moment is to issue two queries like in your final example.

Finally, to help you self-diagnose things like this, I would recommend using the `EXPLAIN` command to see what kind of plan the database generates for each query. You should see that in the subquery case, a spans of `ALL` is generated - that indicates a full table scan.

**Comment by turbo** [comment]
OK, so can I provide an index hint then?

**Comment by RaduBerinde** [comment]
Unfortunately, no - the problem isn't merely that we're not picking the right index, it's that we're not able to generate constraints from subquery results. With the optimizer, this will become a nested-loop join which does what we want (though I'm not sure how the `unnest` will be handled).

**Comment by jordanlewis** [comment]
Closing as duplicate of #23264.
